### PR TITLE
Benchmaking for multi-objectives samplers

### PR DESCRIPTION
**Pull request by drumehiron**
## Motivation
#3245

## Description of the changes

This PR is extend 'optuna/benchmarks/run_kurobako.py' to multi-objectives.

## TODOs:

- [x] fix `--sampler-kwargs-list` argument

### CODE REVIEW

**Comment by HideakiImamura** [comment]
Thanks for the PR! I'd like to investigate how to use `sampler-kwargs-list`.

File: benchmarks/mo_runner.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import sys
+from kurobako import solver
+from kurobako.solver.optuna_multi_objective import OptunaSolverFactory
```
Comment (HideakiImamura) [inline benchmarks/mo_runner.py:3]:
```suggestion
from kurobako.solver.optuna import OptunaSolverFactory
```

File: benchmarks/mo_runner.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+# import json
+
+
+optuna.logging.disable_default_handler()
+
+
+def create_study(n_objectives, seed):
+    directions = ["minimize" for _ in range(n_objectives)]
+
```
Comment (HideakiImamura) [inline benchmarks/mo_runner.py:14]:
```suggestion
n_objectives = 2

```

File: benchmarks/mo_runner.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+from kurobako.solver.optuna_multi_objective import OptunaSolverFactory
+import optuna
+
+# import json
+
+
+optuna.logging.disable_default_handler()
+
+
+def create_study(n_objectives, seed):
```
Comment (HideakiImamura) [inline benchmarks/mo_runner.py:12]:
```suggestion
def create_study(seed: int) -> optuna.Study:
```

File: benchmarks/mo_runner.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    directions = ["minimize" for _ in range(n_objectives)]
+
+    sampler_name = sys.argv[1]
+
+    # Sampler.
+    sampler_cls = getattr(
+        optuna.multi_objective.samplers,
```
Comment (HideakiImamura) [inline benchmarks/mo_runner.py:19]:
```suggestion
        optuna.samplers,
```

File: benchmarks/mo_runner.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    # sampler_kwargs = json.loads(sys.argv[2])
+    # try:
+    #     sampler_kwargs["seed"] = seed
+    #     sampler = sampler_cls(**sampler_kwargs)
+    # except:
+    #     del sampler_kwargs["seed"]
+    #     sampler = sampler_cls(**sampler_kwargs)
+    sampler = sampler_cls()
+
+    return optuna.multi_objective.create_study(directions=directions, sampler=sampler)
```
Comment (HideakiImamura) [inline benchmarks/mo_runner.py:36]:
```suggestion
    return optuna.create_study(
        directions=directions,
        sampler=sampler,
        pruner=optuna.pruners.NopPruner(),
    )
```

File: benchmarks/run_mo_kurobako.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+def run(args: argparse.Namespace) -> None:
+    kurobako_cmd = os.path.join(args.path_to_kurobako, "kurobako")
+    subprocess.run(f"{kurobako_cmd} --version", shell=True)
+
+    if not (os.path.exists(args.data_dir) and os.path.isdir(args.data_dir)):
+        raise ValueError(f"Data directory {args.data_dir} cannot be found.")
+
+    os.makedirs(args.out_dir, exist_ok=True)
+    study_json_fn = os.path.join(args.out_dir, "studies.json")
+    subprocess.check_call(f"echo >| {study_json_fn}", shell=True)
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:15]:
```suggestion
    study_json_filename = os.path.join(args.out_dir, "studies.json")
    subprocess.check_call(f"echo >| {study_json_filename}", shell=True)
```

File: benchmarks/run_mo_kurobako.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    solvers_filename = os.path.join(args.out_dir, "solvers.json")
+    subprocess.check_call(f"echo >| {solvers_filename}", shell=True)
+    problems_filename = os.path.join(args.out_dir, "problems.json")
+    subprocess.check_call(f"echo >| {problems_filename}", shell=True)
+
+    # Create ZDT problems
+    cmd = f"{kurobako_cmd} problem-suite zdt | tee -a {problems_filename}"
+    subprocess.run(cmd, shell=True)
+
+    # Create NAS bench problem(C) (for Multi-Objective Settings).
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:25]:
```suggestion
    # Create NAS bench problem(A) (for Multi-Objective Settings).
```

File: benchmarks/run_mo_kurobako.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    dataset = os.path.join(args.data_dir, "nasbench_full.bin")
+    cmd = (
+        f'{kurobako_cmd} problem nasbench "{dataset}"'
+        f"--encoding C --metrics accuracy params | tee -a {problems_filename}"
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:29]:
```suggestion
        f'{kurobako_cmd} problem nasbench "{dataset}" '
        f"--metrics params accuracy | tee -a {problems_filename}"
```

File: benchmarks/run_mo_kurobako.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        python_command = f"mo_runner.py {sampler} {sampler_kwargs}"
+        cmd = (
+            f"{kurobako_cmd} solver --name {name} command python {python_command}"
+            f"| tee -a {solvers_filename}"
+        )
+        subprocess.run(cmd, shell=True)
+
+    # Create study.
+    cmd = (
+        f"{kurobako_cmd} studies --budget 1000 "
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:54]:
```suggestion
        f"{kurobako_cmd} studies --budget 300 "
```

File: benchmarks/run_mo_kurobako.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        f"--solvers $(cat {solvers_filename}) --problems $(cat {problems_filename}) "
+        f"--repeats {args.n_runs} --seed {args.seed} "
+        f"> {study_json_fn}"
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:57]:
```suggestion
        f"> {study_json_filename}"
```

File: benchmarks/run_mo_kurobako.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    )
+    subprocess.run(cmd, shell=True)
+
+    result_filename = os.path.join(args.out_dir, "results.json")
+    cmd = (
+        f"cat {study_json_fn} | {kurobako_cmd} run --parallelism {args.n_jobs} "
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:63]:
```suggestion
        f"cat {study_json_filename} | {kurobako_cmd} run --parallelism {args.n_jobs} -q "
```

File: benchmarks/run_mo_kurobako.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    subprocess.run(cmd, shell=True)
+
+    # Report
+    report_filename = os.path.join(args.out_dir, "report.md")
+    cmd = f"cat {result_filename} | {kurobako_cmd} report > {report_filename}"
+    subprocess.run(cmd, shell=True)
+
+    # Plot pareto-front.
+    problem_names = ["NASBench", "ZDT1", "ZDT2", "ZDT3", "ZDT4", "ZDT5", "ZDT6"]
+    for problem_name in problem_names:
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:75]:
```suggestion
    xmins = [0, 0, 0, 0, 0, 8, 0.2]
    xmaxs = [25000000, 1, 1, 1, 1, 24, 1]
    ymins = [0, 1, 2, 0, 20, 1, 5]
    ymaxs = [0.2, 7, 7, 7, 250, 6, 10]
    for problem_name, xmin, xmax, ymin, ymax in zip(problem_names, xmins, xmaxs, ymins, ymaxs):
```

File: benchmarks/run_mo_kurobako.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        cmd = (
+            f"cat {result_filename} | grep {problem_name} | "
+            f"{kurobako_cmd} plot pareto-front -o {args.out_dir}"
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:78]:
```suggestion
            f"{kurobako_cmd} plot pareto-front -o {args.out_dir} "
            f"--xmin {xmin} --xmax {xmax} --ymin {ymin} --ymax {ymax}"
```

File: benchmarks/run_mo_kurobako.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--path-to-kurobako", type=str, default="")
+    parser.add_argument("--name-prefix", type=str, default="")
+    parser.add_argument("--n-runs", type=int, default=1)
+    parser.add_argument("--n-jobs", type=int, default=1)
+    parser.add_argument(
+        "--sampler-list",
+        type=str,
+        default="RandomMultiObjectiveSampler NSGAIIMultiObjectiveSampler",
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:92]:
```suggestion
        default="RandomSampler TPESampler NSGAIISampler",
```

File: benchmarks/run_mo_kurobako.py [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    )
+    parser.add_argument("--sampler-kwargs-list", type=str, default="{} {}")
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:94]:
```suggestion
    parser.add_argument("--sampler-kwargs-list", type=str, default="{} {} {}")
```

File: benchmarks/run_mo_kurobako.py (hunk 7 above)
Comment (HideakiImamura) [reply benchmarks/run_mo_kurobako.py:25]: The encoding A is for multi-objective. Sorry for the confusion.

File: benchmarks/run_mo_kurobako.py (hunk 9 above)
Comment (HideakiImamura) [reply benchmarks/run_mo_kurobako.py:54]: The budget 1000 looks too big for TPE (the default sampler of Optuna).

File: benchmarks/run_mo_kurobako.py (hunk 12 above)
Comment (HideakiImamura) [reply benchmarks/run_mo_kurobako.py:75]: Specify a range to make the plot look right.

File: benchmarks/run_mo_kurobako.py (hunk 14 above)
Comment (HideakiImamura) [reply benchmarks/run_mo_kurobako.py:92]: RandomMultiObjectiveSampler and NSGAIIMultiObjectiveSampler are deprecated.

File: benchmarks/mo_runner.py (hunk 4 above)
Comment (HideakiImamura) [reply benchmarks/mo_runner.py:19]: optuna.multi_objective module is deprecated.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR! I made some modifications to make it work with the latest Optuna version. Check it out!

**Comment by HideakiImamura** [comment]
@drumehiron PTAL!

**Comment by drumehiron** [comment]
Sorry for the late reply and thank you for review. 
I check your review.

File: benchmarks/run_mo_kurobako.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+        python_command = f"mo_runner.py {sampler} {sampler_kwargs}"
+        cmd = (
+            f"{kurobako_cmd} solver --name {name} command python {python_command}"
+            f"| tee -a {solvers_filename}"
+        )
+        subprocess.run(cmd, shell=True)
+
+    # Create study.
+    cmd = (
+        f"{kurobako_cmd} studies --budget 300 "
```
Comment (drumehiron) [inline benchmarks/run_mo_kurobako.py:54]: I have confirmed that it works well on my computer with `--budget 10` .
In the case of `-- budget 300`, the computation time is too long, so I haven't checked if it works properly yet.

Comment (HideakiImamura) [reply benchmarks/run_mo_kurobako.py:54]: Agree. How about using the same budget with the existing single-objective one? It is 80. https://github.com/optuna/optuna/blob/master/benchmarks/run_kurobako.py#L69

**Comment by HideakiImamura** [comment]
You can fix the CI error of `Checks` by formatting. 
```
$ bash formats.sh
```

**Comment by drumehiron** [comment]
Sorry for my late response. I'm busy until this week, so I'll check and reply around next Wednesday.

**Comment by hvy** [comment]
For the record https://github.com/optuna/optuna/pull/3349 was recently created. It's a PR based on the changes here. @drumehiron , could you possibly make sure not to squash any changes, to keep the commit history unchanged (and updatable for #3349 )? 🙇

**Comment by drumehiron** [comment]
I see.  Thank you for updating my PR.

**Comment by himkt** [comment]
@drumehiron #3349 is based on this PR (so if this PR changes commit history, #3349 will struggle to catch up) but it doesn't take over your PR. So I think we want you to work on this PR with us. cc. @hvy

**Comment by hvy** [comment]
That's correct. I didn't mean to close this PR. Rather, we'd like to have it merged because #3349 depends on it. Sorry to cause any confusion. Allow me to reopen it.

File: benchmarks/mo_create_study.py [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+from kurobako import solver
+from kurobako.solver.optuna import OptunaSolverFactory
+
+import optuna
+
+
+optuna.logging.disable_default_handler()
+
+
+def create_study(seed: int) -> optuna.Study:
```
Comment (nzw0301) [inline benchmarks/mo_create_study.py:13]: `seed` is not used in this function.

File: benchmarks/run_mo_kurobako.py [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+
+    os.makedirs(args.out_dir, exist_ok=True)
+    study_json_filename = os.path.join(args.out_dir, "studies.json")
+    subprocess.check_call(f"echo >| {study_json_filename}", shell=True)
+    solvers_filename = os.path.join(args.out_dir, "solvers.json")
+    subprocess.check_call(f"echo >| {solvers_filename}", shell=True)
+    problems_filename = os.path.join(args.out_dir, "problems.json")
+    subprocess.check_call(f"echo >| {problems_filename}", shell=True)
+
+    # Create ZDT problems
```
Comment (nzw0301) [inline benchmarks/run_mo_kurobako.py:21]:
```suggestion
    # Create ZDT problems.
```

File: benchmarks/run_mo_kurobako.py [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+    subprocess.run(cmd, shell=True)
+
+    result_filename = os.path.join(args.out_dir, "results.json")
+    cmd = (
+        f"cat {study_json_filename} | {kurobako_cmd} run --parallelism {args.n_jobs} -q "
+        f"> {result_filename}"
+    )
+    subprocess.run(cmd, shell=True)
+
+    # Report
```
Comment (nzw0301) [inline benchmarks/run_mo_kurobako.py:68]:
```suggestion
    # Report.
```

**Review by nzw0301 — COMMENTED** [review]
Let me leave a few minor comments.

**Comment by nzw0301** [comment]
Could you merge the latest master onto the branch? If the branch does not conflict with the master branch, we can use `git rebase`.

**Comment by drumehiron** [comment]
@nzw0301 Thank you for your review. I fix the points.

File: benchmarks/run_mo_kurobako.py [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+    parser.add_argument("--n-jobs", type=int, default=8)
+    parser.add_argument(
+        "--sampler-list",
+        type=str,
+        default="RandomSampler TPESampler NSGAIISampler",
+    )
+    parser.add_argument(
+        "--sampler-kwargs-list",
+        type=str,
+        default=r'{} \{"multivariate":true,"constant_liar":true\} {}',
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:105]: In my environment, I can run with this argument.
```suggestion
        default="{} {\\\"multivariate\\\":true\,\\\"constant_liar\\\":true} {}",
```

File: benchmarks/run_mo_kurobako.py [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+        python_command = f"{args.path_to_create_study} {sampler} {sampler_kwargs}"
+        cmd = (
+            f"{kurobako_cmd} solver --name {name} command python {python_command}"
+            f"| tee -a {solvers_filename}"
+        )
+        subprocess.run(cmd, shell=True)
+
+    # Create study.
+    cmd = (
+        f"{kurobako_cmd} studies --budget 10 "
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:54]: We should use the same value as the single-objective one.
```suggestion
        f"{kurobako_cmd} studies --budget 80 "
```

File: benchmarks/run_mo_kurobako.py [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--path-to-kurobako", type=str, default="")
+    parser.add_argument(
+        "--path-to-create-study", type=str, default="benchmarks/mo_create_study.py"
+    )
+    parser.add_argument("--name-prefix", type=str, default="")
+    parser.add_argument("--n-runs", type=int, default=1)
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:95]: Ditto.
```suggestion
    parser.add_argument("--n-runs", type=int, default=100)
```

File: benchmarks/run_mo_kurobako.py [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+    parser.add_argument("--n-jobs", type=int, default=8)
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:96]: Ditto.
```suggestion
    parser.add_argument("--n-jobs", type=int, default=10)
```

File: benchmarks/run_mo_kurobako.py [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+def run(args: argparse.Namespace) -> None:
+    kurobako_cmd = os.path.join(args.path_to_kurobako, "kurobako")
+    subprocess.run(f"{kurobako_cmd} --version", shell=True)
+
+    if not (os.path.exists(args.data_dir) and os.path.isdir(args.data_dir)):
+        raise ValueError(f"Data directory {args.data_dir} cannot be found.")
+
+    os.makedirs(args.out_dir, exist_ok=True)
+    study_json_filename = os.path.join(args.out_dir, "studies.json")
+    subprocess.check_call(f"echo >| {study_json_filename}", shell=True)
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:15]: Looks good not to use `-n` option since it does not work in macOS.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the update. I have several comment mainly to enable the `sampler-kwargs` option. Please take a look.

File: benchmarks/run_mo_kurobako.py (hunk 24 above)
Comment (drumehiron) [reply benchmarks/run_mo_kurobako.py:15]: Thank you for checking it.

**Review by drumehiron — COMMENTED** [review]
Thank you for review. I checked it.

**Comment by drumehiron** [comment]
I share the graphs obtained by running the program locally. 

<details><summary>Results</summary><div>

## NASBench(A)

nasbench-a-randomsampler-1c7bd31546a52475c4b4f00abd51e68b6fa14baa818299ea2ec51f12703e0f81

nasbench-a-nsgaiisampler-896bca5b8d4113220433e09568af66be3023b7d0b47e741696a856e8f5b2d928

nasbench-a-tpesampler-ffccd9f8c4a9aab456766bd27c3197082fdb1999cc690dda5ac2a539d8f7c0de

## ZDT1

zdt1-randomsampler-86e032ff35d1b5654753829bd6cc20ce1b632aa11603652116c410c244756a7b

zdt1-nsgaiisampler-31990e8998f721570b72be72bb3d479d8d4800e4af3084ad4bc95cac2548ab52

zdt1-tpesampler-3c0bc4817ace50b5fac49c7f9f57f078ca1c0655302582023310f3fc69ad986d

## ZDT2

zdt2-randomsampler-de42e245cdf524596a564993f7f507a985030a6e7aec4203745814c45843c5b6
zdt2-nsgaiisampler-2a2e1baae5b4a6d2965efc144e8b27589b5572b056d5effda7cffdb27a53d392
zdt2-tpesampler-e2ccd295015f6120f4e2fc39a3ff1cf2286036c6fae5b2af8bf88ebe5a7fb1dd

## ZDT3

zdt3-randomsampler-b39795858f4f0ada27a818eb43012e270d54df80975da961484d8583d85e210a
zdt3-nsgaiisampler-d3fbf8e2a0362be1e84132f591dbaf8d24dbf110f228dc8563363623a4f12384
zdt3-tpesampler-c7bf2436cb6b63fa64e83894fe887c5ebc5e279e08a12f4c0b403f8c2bc8f18e

## ZDT4



zdt4-randomsampler-8d80b3b5de53e7beec7dc277ae11f6f59e55d72ef75df2ebff9026b692f88669

zdt4-nsgaiisampler-f04dcae2c25f73a2705d2d6a908bb6fa349ae10ebefa3afdb1b8bbff12eedf68

zdt4-tpesampler-24e89f311b6c0eabe31420949c8a89450825b43550ed0001d97bee7f47c6bfb3

## ZDT5
zdt5-randomsampler-e3c1162ca5533dbca443ecda345a0e0ad6f2977f81f1fe6cdbcad49526cf7303

zdt5-nsgaiisampler-8f790328245e8e701e6e8cd06da1fd99b13aa5ccb063519a2aac62532743fd38

zdt5-tpesampler-8bba257f9b552135e3898209d0d42c16ce9678f4274632c1941d08d79d28d5b5

## ZDT6

zdt6-randomsampler-7c8042f981ed8e6d37b6dfec2d9e0afca4bdcf577a828a0c1c3167abe0aa2315

zdt6-nsgaiisampler-afe417ec4d4b4e6a1332b77a44dbe5b4d7282bb6624c965eb213ad183ad7bf3f

zdt6-tpesampler-0a2d14640e3ca5fafa9802dd461d12a91d944872deb034384ae2a72003a38464

</div></details>

**Comment by HideakiImamura** [comment]
@contramundum53 Could you review this PR if you have time?

File: benchmarks/mo_create_study.py [hunk 25]
```diff
@@ -0,0 +1,44 @@
+import json
```
Comment (HideakiImamura) [inline benchmarks/mo_create_study.py:1]: The `benchmarks` directory becomes a bit complicated. How about creating a `benchmarks/kurobako` directory and moving `mo_create_study.py` into this?

File: benchmarks/mo_create_study.py [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+from kurobako.solver.optuna import OptunaSolverFactory
+
+import optuna
+
+
+optuna.logging.disable_default_handler()
+
+
+def create_study(seed: int) -> optuna.Study:
+    seed
```
Comment (HideakiImamura) [inline benchmarks/mo_create_study.py:14]: Adding a  comment above this line is useful to understand.
```suggestion
    # Avoid the fail by `flake8`.
    seed
```

File: benchmarks/run_mo_kurobako.py [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+    solvers_filename = os.path.join(args.out_dir, "solvers.json")
+    subprocess.check_call(f"echo >| {solvers_filename}", shell=True)
+    problems_filename = os.path.join(args.out_dir, "problems.json")
+    subprocess.check_call(f"echo >| {problems_filename}", shell=True)
```
Comment (HideakiImamura) [inline benchmarks/run_mo_kurobako.py:19]: Can be simplified as #3468 do so.
```suggestion
    solvers_filename = os.path.join(args.out_dir, "solvers.json")
    problems_filename = os.path.join(args.out_dir, "problems.json")
    
    # Ensure all files are empty.
    for filename in [study_json_filename, solvers_filename, problems_filename]:
        with open(filename, "w"):
            pass
```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the update! Almost, LGTM. I have several minor comments. PTAL.

File: benchmarks/run_mo_kurobako.py [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+        python_command = f"{args.path_to_create_study} {sampler} {sampler_kwargs}"
+        cmd = (
+            f"{kurobako_cmd} solver --name {name} command python {python_command}"
+            f"| tee -a {solvers_filename}"
+        )
+        subprocess.run(cmd, shell=True)
+
+    # Create study.
+    cmd = (
+        f"{kurobako_cmd} studies --budget 80 "
```
Comment (contramundum53) [inline benchmarks/run_mo_kurobako.py:54]:
```suggestion
        f"{kurobako_cmd} studies --budget 120 "
```
I think we need more iterations (generations) to test NSGA-II effectively.

File: benchmarks/run_mo_kurobako.py [hunk 29]
```diff
@@ -0,0 +1,112 @@
...
+    parser.add_argument("--n-jobs", type=int, default=10)
+    parser.add_argument(
+        "--sampler-list",
+        type=str,
+        default="RandomSampler TPESampler NSGAIISampler",
+    )
+    parser.add_argument(
+        "--sampler-kwargs-list",
+        type=str,
+        default=r"{} {\"multivariate\":true\,\"constant_liar\":true} {}",
```
Comment (contramundum53) [inline benchmarks/run_mo_kurobako.py:105]:
```suggestion
        default=r"{} {\"multivariate\":true\,\"constant_liar\":true} {\"population_size\":20}",
```
By default, the population size for NSGA-II is 50, but I think we need more generations for meaningful benchmark.

**Review by contramundum53 — COMMENTED** [review]
Thanks for your PR! It looks almost good to me. I left some minor comments.

File: benchmarks/mo_create_study.py (hunk 25 above)
Comment (drumehiron) [reply benchmarks/mo_create_study.py:1]: I agree your comments. I move `mo_create_study.py` into `into benchmarks/kurobako`.

File: benchmarks/mo_create_study.py (hunk 26 above)
Comment (drumehiron) [reply benchmarks/mo_create_study.py:14]: LGTM. I add this comment.

File: benchmarks/run_mo_kurobako.py [hunk 30]
```diff
@@ -0,0 +1,112 @@
...
+        raise ValueError(
+            "The number of samplers does not match the given keyword arguments. \n"
+            f"sampler_list: {sampler_list}, sampler_kwargs_list: {sampler_kwargs_list}."
+        )
+
+    for sampler, sampler_kwargs in zip(sampler_list, sampler_kwargs_list):
+        name = f"{args.name_prefix}_{sampler}"
+        python_command = f"{args.path_to_create_study} {sampler} {sampler_kwargs}"
+        cmd = (
+            f"{kurobako_cmd} solver --name {name} command python {python_command}"
```
[the remark this answers is not in the document]
Comment (drumehiron) [reply benchmarks/run_mo_kurobako.py:47]: Thank you for review, I change `python` to `python3`

File: benchmarks/run_mo_kurobako.py (hunk 29 above)
Comment (drumehiron) [reply benchmarks/run_mo_kurobako.py:105]: Thank you for comments.  I change budget and population size. I will share results on NSGA-II.

**Review by drumehiron — COMMENTED** [review]
Thank you for review. I will update this PR.

**Comment by drumehiron** [comment]
I share the graphs obtained by running the program locally.  NSGA-II seems to be behaving better.

<details/><div/>

## NAS Bench

nasbench-a-randomsampler-1b4ee96b52429f91f3d568cb495d6aefc737bd7283867abd7dbf4e3f84ee2636

nasbench-a-nsgaiisampler-43161816893134c184b8a635cd049cfc24970811659b0e55c99861b71a9f9786

nasbench-a-tpesampler-e793c22f63b87027c2a37820263cab7ade42dd172c7abccd759a7f7f4ddfe775

## ZDT1

zdt1-randomsampler-c172753ed599a19d9ee04801069bf07724f21af549892849062aeff8fa78ed04

zdt1-nsgaiisampler-062702bb0530d691c615c3d79402539bf4e0792e4ab6077033a1669aef1c5960

zdt1-tpesampler-ee52859e8ee615ef9ac3ac1f4e3067a9cb864a84f02c918d61d9bbd4290313b8

## ZDT2

zdt2-randomsampler-0cc7b8fe53a63b12599b00b78a042204da264bb705e1f692c2254a534b28d8bd

zdt2-nsgaiisampler-37c6c3d41cfdbfd5c874e53cfcdeaccec86b7540cdd9c6ce60b85d84054f5cf3

zdt2-tpesampler-e2ba32b9808b0058363379c51ad4fc778b47f40750b9eda06b47e29b01b1da18

## ZDT3

zdt3-randomsampler-84cafeeda7ed51e572324a219c9aab46b26ecf37a07a35295867ae8a2a4387d5
zdt3-nsgaiisampler-bbdccb93ce6ebf5860610e6ce1e78c949754acf3299adcfd65e7dd738ab74c5e
zdt3-tpesampler-2de1fca7d592bdceffa6e145694b7b70aa6dcb94e807589a7eabe2a1d974653f

## ZDT4

zdt4-randomsampler-2a8a86bbd7ec974913f37e1a8d2ad1f7224473fd36526faa57dd601addbf46fe

zdt4-nsgaiisampler-850c9e57c9e7b2e0ae8db0b443c350e359e33a476082208e86be88d307a02e7c

zdt4-tpesampler-ddac64ba539ced4764fe5561cbb34f400408e2129e3efa54e94e8f3e93aae475

## ZDT5

zdt5-randomsampler-11f3701d36d8cdf6c370aa2ed262e5cf345d4b49fb344109336bccd684febef4

zdt5-nsgaiisampler-cb0cda6329716454d1edaafecad98d2e7bd758647b78148eed26f7ff9a50b27d

zdt5-tpesampler-29c17cff0a3e3223a0bb53c5dd37f2a7ec57436384f3aa34ba8187d9331d7b01

## ZDT6

zdt6-randomsampler-2aba1d35972c6e95106e4b4550f6e41f7afd0ec085a5db2c11edb46934d28f97

zdt6-nsgaiisampler-0db755f627a57014946365bf9a8ad1ee79550cd4e666dd8ce769173558c0e493


zdt6-tpesampler-94cf5a38e7617ff6cbdffb4add11aa4e3cdb0a9ea2718af69b64b9cc3c9cee05

</div></details>

**Review by contramundum53 — APPROVED** [review]
Thank you!

**Comment by contramundum53** [comment]
Maybe we should also change the single-objective version to have budget 120 for consistency, but I think that can be another PR.

**Pull request merged**